File: src/main/kotlin/fr/ignishky/framework/cqrs/command/middleware/EventPersistenceMiddleware.java
package fr.ignishky.framework.cqrs.command.middleware;

import fr.ignishky.framework.cqrs.command.Command;
import fr.ignishky.framework.cqrs.event.Event;
import fr.ignishky.framework.cqrs.event.EventStore;
import fr.ignishky.framework.domain.CorrelationId;

import java.util.List;

public class EventPersistenceMiddleware extends CommandMiddleware {

    private final EventStore eventStore;

    public EventPersistenceMiddleware(CommandMiddleware next, EventStore eventStore) {
        super(next);
        this.eventStore = eventStore;
    }

    @Override
    public List<Event<?, ?, ?>> handle(Command command, CorrelationId correlationId) {
        var response = next(command, correlationId);
        eventStore.store(response);
        return response;
    }

    public record Builder(EventStore eventStore) implements CommandMiddlewareBuilder {

        @Override
        public CommandMiddleware chain(CommandMiddleware next) {
            return new EventPersistenceMiddleware(next, eventStore);
        }

    }

}

File: src/main/kotlin/fr/ignishky/framework/cqrs/command/middleware/EventPersistenceMiddleware.kt
package fr.ignishky.framework.cqrs.command.middleware

import com.fasterxml.jackson.databind.ObjectMapper
import fr.ignishky.framework.cqrs.command.Command
import fr.ignishky.framework.cqrs.event.Event
import fr.ignishky.framework.cqrs.event.spi.postgres.EventEntity
import fr.ignishky.framework.cqrs.event.spi.postgres.EventRepository
import fr.ignishky.framework.domain.CorrelationId

class EventPersistenceMiddleware(
    next: CommandMiddleware,
    private val objectMapper: ObjectMapper,
    private val eventRepository: EventRepository
) : CommandMiddleware(next) {

    override fun handle(command: Command, correlationId: CorrelationId): List<Event<*, *, *>> {
        val events = next(command, correlationId)
        eventRepository.saveAll(events.map {
            EventEntity(
                0,
                it.aggregateId.value(),
                it.aggregateClass.simpleName.toString(),
                it::class.simpleName.toString(),
                it.instant,
                objectMapper.writeValueAsString(it.payload),
                correlationId.value
            )
        })
        return events
    }

    class Builder(private val objectMapper: ObjectMapper, private val eventRepository: EventRepository) : CommandMiddlewareBuilder {

        override fun chain(next: CommandMiddleware): CommandMiddleware {
            return EventPersistenceMiddleware(next, objectMapper, eventRepository)
        }
    }

}
